rueidisotel/metrics.go

```diff
@@ -0,0 +1,163 @@
+func NewClient(opts ...Option) (rueidis.Client, error) {
+	oclient := newClient(opts...)
+
+	attempt, err := oclient.meter.Int64Counter("rueidis_dial_attempt")
+	if err != nil {
+		return nil, err
+	}
+	oclient.attempt = attempt
+
+	success, err := oclient.meter.Int64Counter("rueidis_dial_success")
+	if err != nil {
+		return nil, err
+	}
+	oclient.success = success
+
+	conns, err := oclient.meter.Int64UpDownCounter("rueidis_dial_conns")
+	if err != nil {
+		return nil, err
+	}
+	oclient.conns = conns
+
+	dialLatency, err := oclient.meter.Float64Histogram(
+		"rueidis_dial_latency",
+		metric.WithUnit("s"),
+		metric.WithExplicitBucketBoundaries(oclient.dialLatencyHistogramOption.Buckets...),
+	)
+	if err != nil {
+		return nil, err
+	}
+	oclient.dialLatency = dialLatency
+
+	oclient.clientOption.DialFn = trackDialing(
+		attempt, success, conns, dialLatency, oclient.clientOption.DialFn,
+	)
+	cli, err := rueidis.NewClient(oclient.clientOption)
+	if err != nil {
+		return nil, err
+	}
+	oclient.client = cli
+
+	return oclient, nil
+}
+
+func newClient(opts ...Option) *otelclient {
+	cli := &otelclient{}
+	for _, opt := range opts {
+		opt(cli)
+	}
+	if cli.clientOption.DialFn == nil {
+		cli.clientOption.DialFn = DefaultDialFn
+	}
+	if cli.dialLatencyHistogramOption.Buckets == nil {
+		cli.dialLatencyHistogramOption.Buckets = DefaultDialLatencyHistogramDefaultBuckets
+	}
+	if cli.meterProvider == nil {
+		cli.meterProvider = otel.GetMeterProvider() // Default to global MeterProvider
+	}
+	if cli.tracerProvider == nil {
+		cli.tracerProvider = otel.GetTracerProvider() // Default to global TracerProvider
+	}
+
+	// Now that we have the meterProvider and tracerProvider, get the Meter and Tracer
+	cli.meter = cli.meterProvider.Meter(name)
+	cli.tracer = cli.tracerProvider.Tracer(name)
+	// Now create the counters using the meter
+	cli.cscMiss, _ = cli.meter.Int64Counter("rueidis_do_cache_miss")
+	cli.cscHits, _ = cli.meter.Int64Counter("rueidis_do_cache_hits")
+	return cli
+}
+
+func trackDialing(
+	attempt metric.Int64Counter,
+	success metric.Int64Counter,
+	conns metric.Int64UpDownCounter,
+	dialLatency metric.Float64Histogram,
+	dialFn func(string, *net.Dialer, *tls.Config) (conn net.Conn, err error),
+) func(string, *net.Dialer, *tls.Config) (conn net.Conn, err error) {
+	return func(network string, dialer *net.Dialer, tlsConfig *tls.Config) (conn net.Conn, err error) {
+		ctx := context.Background()
+		attempt.Add(ctx, 1)
+
+		start := time.Now()
+
+		conn, err = dialFn(network, dialer, tlsConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		dialLatency.Record(ctx, time.Since(start).Seconds())
+		success.Add(ctx, 1)
+		conns.Add(ctx, 1)
+
+		return &connTracker{
+			Conn:  conn,
+			conns: conns,
+			once:  0,
+		}, nil
+	}
+}
+
+type connTracker struct {
+	net.Conn
+	conns metric.Int64UpDownCounter
+	once  int32
+}
+
+func (t *connTracker) Close() error {
+	v := atomic.AddInt32(&t.once, 1)
+	if v == 1 {
```

Comment:
```suggestion
	if atomic.CompareAndSwapInt32(&t.once, 0, 1) {
```

Even if it is hard to overflow the `t.once`, using `CompareAndSwapInt32` is still better.

Reply:
[a2b467f](https://github.com/redis/rueidis/pull/421/commits/a2b467ff9c8a0de0dc0dbfca1e2d3646659f6c1c)
Oh, yes. definitely better.